samza-core/src/main/java/org/apache/samza/table/remote/TableReadFunction.java

```diff
@@ -44,22 +50,54 @@
 public interface TableReadFunction<K, V> extends Serializable, InitableFunction, ClosableFunction {
   /**
    * Fetch single table record for a specified {@code key}. This method must be thread-safe.
+   * The default implementation calls getAsync and blocks on the completion afterwards.
    * @param key key for the table record
    * @return table record for the specified {@code key}
    */
-  V get(K key);
+  default V get(K key) {
+    try {
+      return getAsync(key).get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new SamzaException("GET failed for " + key, e);
+    }
+  }
+
+  /**
+   * Asynchronously fetch single table record for a specified {@code key}. This method must be thread-safe.
+   * @param key key for the table record
+   * @return CompletableFuture for the get request
+   */
+  CompletableFuture<V> getAsync(K key);
 
   /**
    * Fetch the table {@code records} for specified {@code keys}. This method must be thread-safe.
+   * The default implementation calls getAllAsync and blocks on the completion afterwards.
    * @param keys keys for the table records
-   * @return all records for the specified keys if succeeded; depending on the implementation
-   * of {@link TableReadFunction#get(Object)} it either returns records for a subset of the
-   * keys or throws exception when there is any failure.
+   * @return all records for the specified keys.
    */
   default Map<K, V> getAll(Collection<K> keys) {
-    Map<K, V> records = new HashMap<>();
-    keys.forEach(k -> records.put(k, get(k)));
-    return records;
+    try {
+      return getAllAsync(keys).get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new SamzaException("GET_ALL failed for " + keys, e);
+    }
+  }
+
+  /**
+   * Asynchronously fetch the table {@code records} for specified {@code keys}. This method must be thread-safe.
+   * The default implementation calls getAsync for each key and return a combined future.
+   * @param keys keys for the table records
+   * @return CompletableFuture for the get request
+   */
+  default CompletableFuture<Map<K, V>> getAllAsync(Collection<K> keys) {
+    // Call getAsync() for each key and returned the combined future
+    Map<CompletableFuture<V>, K> getFutures = keys.stream()
+        .collect(Collectors.toMap(k -> getAsync(k), Function.identity()));
+    return CompletableFuture.allOf(
+        Iterables.toArray(getFutures.keySet(), CompletableFuture.class))
+        .thenApply(future ->
+            getFutures.keySet().stream()
+              .collect(Collectors.toMap(f -> getFutures.get(f), f -> f.join())));
```

Comment:
consider restructuring as mentioned earlier? We shouldn't have to use reverse-index the future to its associated key. `map<future<v>, k>`